pkg/domain/stateDiffComponent.go

```diff
@@ -0,0 +1,187 @@
+		diff.InstallationState,
+	)
+}
+
+// determineComponentDiffs creates ComponentDiffs for all components in the blueprint and all installed components as well.
+func determineComponentDiffs(logger logr.Logger, blueprintComponents []Component, installedComponents map[string]*ecosystem.ComponentInstallation) []ComponentDiff {
+	var componentDiffs = map[string]ComponentDiff{}
+	for _, blueprintComponent := range blueprintComponents {
+		installedComponent := installedComponents[blueprintComponent.Name]
+		componentDiffs[blueprintComponent.Name] = determineComponentDiff(logger, &blueprintComponent, installedComponent)
+	}
+
+	for _, installedComponent := range installedComponents {
+		blueprintComponent, notFound := findComponentByName(blueprintComponents, installedComponent.Name)
+
+		if notFound == nil {
+			componentDiffs[installedComponent.Name] = determineComponentDiff(logger, &blueprintComponent, installedComponent)
+			continue
+		}
+
+		var notFoundInBlueprint *Component = nil
+		componentDiffs[installedComponent.Name] = determineComponentDiff(logger, notFoundInBlueprint, installedComponent)
+	}
+	return maps.Values(componentDiffs)
+}
+
+// determineComponentDiff creates a ComponentDiff out of a Component from the blueprint and the ecosystem.ComponentInstallation in the ecosystem.
+// If the Component is nil (was not in the blueprint), the actual state is also the expected state.
+// If the installedComponent is nil, it is considered to be not installed currently.
+func determineComponentDiff(logger logr.Logger, blueprintComponent *Component, installedComponent *ecosystem.ComponentInstallation) ComponentDiff {
+	var expectedState, actualState ComponentDiffState
+	componentName := "" // either blueprintComponent or installedComponent could be nil
+
+	if installedComponent == nil {
+		actualState = ComponentDiffState{
+			InstallationState: TargetStateAbsent,
+		}
+	} else {
+		componentName = installedComponent.Name
+		actualState = ComponentDiffState{
+			DistributionNamespace: installedComponent.DistributionNamespace,
+			Version:               installedComponent.Version,
+			InstallationState:     TargetStatePresent,
+		}
+	}
+
+	if blueprintComponent == nil {
+		expectedState = actualState
+	} else {
+		componentName = blueprintComponent.Name
+		expectedState = ComponentDiffState{
+			DistributionNamespace: blueprintComponent.DistributionNamespace,
+			Version:               blueprintComponent.Version,
+			InstallationState:     blueprintComponent.TargetState,
+		}
+	}
+
+	return ComponentDiff{
+		Name:         componentName,
+		Expected:     expectedState,
+		Actual:       actualState,
+		NeededAction: getNextComponentAction(logger, expectedState, actualState),
+	}
+}
+
+func findComponentByName(components []Component, name string) (Component, error) {
+	for _, component := range components {
+		if component.Name == name {
+			return component, nil
+		}
+	}
+	return Component{}, fmt.Errorf("could not find component '%s'", name)
+}
```

Comment:
```suggestion
func findComponentByName(components []Component, name string) (Component, bool) {
	for _, component := range components {
		if component.Name == name {
			return component, true
		}
	}
	return Component{}, false
}
```

I think a simple boolean value is much better suited here, since we don't use the actual error value anyway.

Reply:
Nice catch.
- [x] done